### PR TITLE
🏗 [test-case-reporting] POST test results from Travis using a gulp task

### PR DESCRIPTION
**Pull request by Rafer45**
/cc @rcebulko, @estherkim
Closes ampproject/amp-github-apps#891
Works toward https://github.com/ampproject/amp-github-apps/issues/922
Replaces https://github.com/ampproject/amphtml/pull/29024 for a cleaner PR history.

### CODE REVIEW

File: build-system/pr-check/checks.js [hunk 1]
```diff
@@ -59,6 +59,7 @@ async function main() {
     timedExecOrDie('gulp dep-check');
     timedExecOrDie('gulp check-types');
     timedExecOrDie('gulp check-sourcemaps');
+    timedExecOrDie('gulp report-test-results');
```
Comment (Rafer45) [inline build-system/pr-check/checks.js:62]: This is very much not the right file for this; I was following the owners-check task too literally. I'll move this to somewhere more suitable.

File: build-system/tasks/report-test-results.js [hunk 2]
```diff
@@ -20,11 +20,78 @@
...
+} = require('../common/travis');
+const {cyan, green, yellow} = require('ansi-colors');
+
+const REPORTING_API_URL = 'https://amp-test-cases.appspot.com/report';
+
+async function getReport(testType) {
+  const karmaReportString = await fs.readFile(
+    `result-reports/${testType}.json`
+  );
+  const karmaReportJson = JSON.parse(karmaReportString);
```
Comment (rcebulko) [inline build-system/tasks/report-test-results.js:40]: nit: this is one of very few situations where using `then` in conjunction with `await` could actually work nicely, ie.

```javascript
async function getReport(testType) {
  const report = await fs.readFile(`result-reports/${testType}.json`)
      .then(JSON.parse);
  return addJobAndBuildInfo(testType, report);
}
```

File: build-system/tasks/report-test-results.js [hunk 3]
```diff
@@ -20,11 +20,78 @@
...
+      cyan(REPORTING_API_URL)
+    );
+  } else {
+    log(
+      yellow('WARNING:'),
+      'failed to report results of type',
+      cyan(testType),
+      ': \n',
+      yellow(response.statusText).substr(0, 100)
+    );
```
Comment (rcebulko) [inline build-system/tasks/report-test-results.js:85]: Good job on keeping the logging conventions!

File: build-system/pr-check/checks.js (hunk 1 above)
Comment (Rafer45) [reply build-system/pr-check/checks.js]: Done.

File: build-system/tasks/report-test-results.js [hunk 4]
```diff
@@ -20,11 +20,78 @@
...
+      'reported to',
+      cyan(REPORTING_API_URL)
+    );
+  } else {
+    log(
+      yellow('WARNING:'),
+      'failed to report results of type',
+      cyan(testType),
+      ': \n',
+      yellow(response.statusText).substr(0, 100)
```
Comment (rcebulko) [inline build-system/tasks/report-test-results.js:84]: The color functions work by wrapping text in [ANSI color codes](https://en.wikipedia.org/wiki/ANSI_escape_code#Colors), like `ESC[31m`. There are "open" and "close" tags, so by substring-ing the colorized contents, you may be cutting off the closing color tag. You probably want to substring the `statusText`.

On that note, why substring at all? AFAIK it's never more than a couple of words

File: build-system/tasks/report-test-results.js [hunk 5]
```diff
@@ -20,11 +20,78 @@
...
+  }
+}
 
 async function reportTestResults() {
-  log('Hello,', green('gulped'), 'world');
+  const filenames = await fs.readdir('result-reports/');
+  const testTypes = filenames.map((filename) => path.parse(filename).name);
+  for (const testType of testTypes) {
+    sendTravisKarmaReport(testType);
+  }
```
Comment (rcebulko) [inline build-system/tasks/report-test-results.js:94]: `await Promise.all(testTypes.map(sendTravisKarmaReport));`

File: build-system/tasks/report-test-results.js (hunk 2 above)
Comment (rcebulko) [reply build-system/tasks/report-test-results.js:40]: Addendum: `require` also natively supports JSON and caches across invocations, so I wouldn't be opposed to the following:

```javascript
async function getReport(testType) {
  return addJobAndBuildInfo(
    testType,
    require(`result-reports/${testType}.json`)
  );
}
```

File: build-system/tasks/report-test-results.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  travisJobNumber,
+  travisCommitSha,
+} = require('../common/travis');
+const {cyan, green, yellow} = require('ansi-colors');
+
+const REPORTING_API_URL = 'https://amp-test-cases.appspot.com/report';
+
+async function getReport(testType) {
+  const karmaReportString = await fs.readFile(
+    `result-reports/${testType}.json`
```
Comment (Rafer45) [inline build-system/tasks/report-test-results.js:38]: Hm, on second thought, I think a helper for the target directory name would be better. We _will_ have to import the function in runtime-test-base.js; however, modularity is not harmed by doing that import -- modularity was already harmed by the fact that the file names need to agree in both places.

Put another way, having a raw `result-reports/${config.testType}.json` in runtime-test-base.js isn't more or less modular than making a helper function and importing it there, because the end result is about the same -- both files have to agree on the filename, so they're (a little) coupled in that way.

It may be better to have the helper method in a separate file so that this file is just in charge of the gulp task itself; it may also be better to just not worry about it, since the filename is used in like 3 or 4 places.

Thoughts? @rcebulko @estherkim

Comment (rcebulko) [reply build-system/tasks/report-test-results.js:38]: Unfortunately some build-system code just relies on the location of some sets of files. At most I could see exporting a constant `RESULT_REPORT_DIRECTORY` to be used elsewhere, but I'll defer entirely to Esther if she disagrees as I have less familiarity with the conventions in build-system.

File: build-system/tasks/report-test-results.js (hunk 2 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js:40]: Oh, wow, I didn't know require worked for that. I think I prefer the first version, but if you feel strongly about it I'll do the second one.

File: build-system/tasks/report-test-results.js (hunk 3 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js:85]: Thank you!

File: build-system/tasks/report-test-results.js (hunk 4 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js:84]: I took the idea from https://github.com/ampproject/amphtml/blob/f8ee003f7239be737c03a6626b0a326c5e9aae6a/build-system/tasks/report-test-status.js -- I figured it would be better to limit the length of the response just in case. I'll remove it.

File: build-system/tasks/report-test-results.js (hunk 6 above)
Comment (estherkim) [reply build-system/tasks/report-test-results.js:38]: Let's leave it as is for now. Generally I like in-place strings more than exported consts because it's way easier to read.

But if it makes sense to have it all in 1 place, we can optimize later.

File: build-system/tasks/report-test-results.js (hunk 5 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js]: Done.

File: build-system/tasks/report-test-results.js (hunk 4 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js]: Done removing it (if you think it'd be better to keep it, we can add it to statusText like you suggested).

File: build-system/tasks/report-test-results.js (hunk 2 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js]: Done implementing the first version

File: build-system/tasks/report-test-results.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  const report = await fs
+    .readFile(`result-reports/${testType}.json`)
+    .then(JSON.parse);
+
+  return addJobAndBuildInfo(testType, report);
+}
+
+function addJobAndBuildInfo(testType, reportJson) {
+  const build = {
+    buildNumber: travisBuildNumber(),
```
Comment (Rafer45) [inline build-system/tasks/report-test-results.js:46]: Should we throw an error if there isn't a travis build number (e.g. if this is called outside a travis build)?

Comment (rcebulko) [reply build-system/tasks/report-test-results.js:46]: WDYT?

Comment (Rafer45) [reply build-system/tasks/report-test-results.js:46]: I think yes; I don't believe this should be getting called outside of builds at all.

File: build-system/pr-check/local-tests.js [hunk 8]
```diff
@@ -45,6 +45,7 @@ function main() {
     timedExecOrDie('gulp integration --nobuild --headless --coverage');
     timedExecOrDie('gulp unit --nobuild --headless --coverage');
     timedExecOrDie('gulp codecov-upload');
+    timedExecOrDie('gulp report-test-results');
```
Comment (estherkim) [inline build-system/pr-check/local-tests.js:48]: Let's do this in a separate PR, including adding the `--report` flag.

File: gulpfile.js [hunk 9]
```diff
@@ -169,6 +170,7 @@ createTask('prettify', prettify);
 createTask('process-3p-github-pr', process3pGithubPr);
 createTask('process-github-issues', processGithubIssues);
 createTask('release', release);
+createTask('report-test-results', reportTestResults);
```
Comment (estherkim) [inline gulpfile.js:173]: nit: can we rename this to something-something-upload like `codecov-upload`?

File: build-system/pr-check/local-tests.js (hunk 8 above)
Comment (Rafer45) [reply build-system/pr-check/local-tests.js]: Done.

File: gulpfile.js (hunk 9 above)
Comment (Rafer45) [reply gulpfile.js]: Done.

File: build-system/tasks/report-test-results.js (hunk 7 above)
Comment (Rafer45) [reply build-system/tasks/report-test-results.js]: Added exceptions for when any of the `travisSomethingSomething` calls fail, let me know what you think.

File: build-system/tasks/report-upload.js [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  const testTypes = filenames.map((filename) => path.parse(filename).name);
+
+  await Promise.all(testTypes.map(sendTravisKarmaReport));
+}
+
+module.exports = {
+  reportUpload,
+};
+
+reportUpload.description = 'Sends test results to test result database';
```
Comment (rcebulko) [inline build-system/tasks/report-upload.js:106]: This file LGTM, let's add some JSDoc comments to at least the main/exported function though

**Comment by Rafer45** [comment]
> How about the name `test-report-upload` or `results-upload`, since "report" can be pretty generic

SGTM; I'll get on that.

File: build-system/tasks/report-upload.js (hunk 10 above)
Comment (Rafer45) [reply build-system/tasks/report-upload.js]: Done.
